### PR TITLE
Implement Beat Up

### PR DESCRIPTION
**Pull request by happinyz**
This PR implements the move Beat Up. It's a pretty unique move with mechanics not found anywhere so a lot of this code is tightly coupled together.

[How Beat Up works](https://pokemondb.net/move/beat-up)

https://github.com/pagefaultgames/pokerogue/assets/59787978/fd0c4692-1b64-4488-ab5a-9ffc58a33e93

### CODE REVIEW

**Comment by happinyz** [comment]
@bennybroseph @Madmadness65 Could I get a review on this? Thanks!

**Review by bennybroseph — CHANGES_REQUESTED** [review]
Needs TSDoc documentation on all new classes and functions and code comments where applicable.

**Comment by happinyz** [comment]
@bennybroseph Could you take another look at this? thanks!

File: src/data/move.ts [hunk 1]
```diff
@@ -1866,6 +1872,45 @@ export class MovePowerMultiplierAttr extends VariablePowerAttr {
...
+ * @returns The base power of the Beat Up hit.
+ */
+const beatUpFunc = (user: Pokemon, allyIndex: number): number => {
+  const party = user.scene.getParty();
+
+  for (let i = allyIndex; i < party.length; i++) {
+    const pokemon = party[i];
+
+    // The user contributes to Beat Up regardless of status condition.
+    // Allies can contribute only if they do not have a non-volatile status condition.
```
Comment (bennybroseph) [inline src/data/move.ts:1888]: this is referenced in two places, is this by mistake?

File: src/data/move.ts [hunk 2]
```diff
@@ -1069,6 +1070,11 @@ export class MultiHitAttr extends MoveAttr {
             hitTimes = 10;
         }
         break;
+      case MultiHitType.BEAT_UP:
+        // No status means the ally pokemon can contribute to Beat Up
+        hitTimes = user.scene.getParty().reduce((total, pokemon) => {
+          return total + (pokemon.id === user.id ? 1 : pokemon?.status && pokemon.status.effect !== StatusEffect.NONE ? 0 : 1)
+        }, 0);
```
Comment (bennybroseph) [inline src/data/move.ts:1077]: here as well

**Review by bennybroseph — CHANGES_REQUESTED** [review]

File: src/data/move.ts (hunk 1 above)
Comment (happinyz) [reply src/data/move.ts:1888]: No, we do this in line 1075 to determine how many times Beat Up can hit, then do it again here to determine which party member's turn it is to hit

**Review by bennybroseph — APPROVED** [review]

**Pull request merged**